doc/case-studies/xycloans/MCxycloans_monitor.ejs.tla

```diff
@@ -0,0 +1,167 @@
+    <%-
+    balanceAddrs
+        .map((addr) => `    "${addr}"`)
+        .join(",\n    ")
+    %>
+}
+
+\* addresses that hold token balances
+TOKEN_ADDR == {
+    <%
+    const tokenAddrs =
+        Object.keys(storage[storage[contractId].instance.TokenId].persistent)
+            .filter((key) => key.startsWith("Balance,"))
+            .filter((key) => key !== `Balance,${contractId}`)
+            .map((key) => key.split(",")[1])
+    %>
+    <%-
+    tokenAddrs
+        .map((addr) => `    "${addr}"`)
+        .join(",\n    ")
+    %>
+}
+
+\* the pool of addresses to draw the values from
+ADDR == { XYCLOANS, XLM_TOKEN_SAC_TESTNET } \union TOKEN_ADDR \union USER_ADDR
+
+VARIABLES
+    \* @type: $tx;
+    last_tx,
+    \* @type: Str -> Int;
+    shares,
+    \* @type: Int;
+    total_shares,
+    \* @type: Int;
+    fee_per_share_universal,
+    \* Keep track of the current storage,
+    \* which can be only changed by a successful transaction.
+    \* @type: $storage;
+    storage
+
+INSTANCE xycloans_monitor
+
+Init ==
+    LET init_stor == [
+        self_instance |-> [
+            FeePerShareUniversal |-> <%- storage[contractId].instance.FeePerShareUniversal %>,
+            TokenId |-> "<%- storage[contractId].instance.TokenId %>"
+        ],
+        self_persistent |-> [
+            Balance |-> SetAsFun({<%-
+              Object.keys(storage[contractId].persistent)
+                .filter((key) => key.startsWith("Balance,"))
+                .map((key) => key.split(",")[1])
+                .map((addr) => `<<"${addr}", ${storage[contractId].persistent["Balance," + addr]}>>`)
+                .join(",\n              ")
+            %>}),
+            MaturedFeesParticular |-> SetAsFun({<%-
+              Object.keys(storage[contractId].persistent)
+                .filter((key) => key.startsWith("MaturedFeesParticular,"))
+                .map((key) => key.split(",")[1])
+                .map((addr) => `<<"${addr}", ${storage[contractId].persistent["MaturedFeesParticular," + addr]}>>`)
+                .join(",\n              ")
+            %>}),
+            FeePerShareParticular |-> SetAsFun({<%-
+              Object.keys(storage[contractId].persistent)
+                .filter((key) => key.startsWith("FeePerShareParticular,"))
+                .map((key) => key.split(",")[1])
+                .map((addr) => `<<"${addr}", ${storage[contractId].persistent["FeePerShareParticular," + addr]}>>`)
+                .join(",\n              ")
+            %>})
+        ],
+        token_persistent |-> [ Balance |-> SetAsFun({<%-
+              Object.keys(storage[storage[contractId].instance.TokenId].persistent)
+                .filter((key) => key.startsWith("Balance,"))
+                .map((key) => key.split(",")[1])
+                .map((addr) =>
+                  `<<"${addr}", ${storage[storage[contractId].instance.TokenId].persistent["Balance," + addr].amount}>>`)
+                .join(",\n              ")
+        %>})]
```

Comment:
Can we extract this into some kind of helper?
It looks like we need this pattern again and again.

Reply:
Good point. I have found a way to introduce definitions in EJS